File: projects/samples/devices/controllers/distance_sensor/distance_sensor.py
# Copyright 1996-2023 Cyberbotics Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
An example of use of the distance sensors.
"""

from controller import Robot


class Controller(Robot):
    NB_SENSORS = 8
    RANGE = 512
    MATRIX = [[11, 12, 8, -2, -3, -5, -7, -9], [-9, -8, -5, -1, -2, 6, 12, 11]]
    timeStep = 64

    def __init__(self):
        super(Controller, self).__init__()
        self.ps = []
        for i in range(self.NB_SENSORS):
            self.ps.append(self.getDevice('ds' + str(i)))
            self.ps[i].enable(self.timeStep)

        self.left_motor = self.getDevice('left wheel motor')
        self.right_motor = self.getDevice('right wheel motor')
        self.left_motor.setPosition(float('inf'))
        self.right_motor.setPosition(float('inf'))
        self.left_motor.setVelocity(0.5)
        self.right_motor.setVelocity(-0.5)

    def run(self):
        while self.step(self.timeStep) != -1:
            sensor_value = []
            for i in range(self.NB_SENSORS):
                sensor_value.append(self.ps[i].getValue())
            speed = [0, 0]
            for i in range(2):
                for j in range(self.NB_SENSORS):
                    # We need to recenter the value of the sensor to be able to get
                    # negative values too. This will allow the wheels to go  backward too.
                    speed[i] += self.MATRIX[i][j] * (1 - (sensor_value[j] / self.RANGE))

            # Set the motor speeds
            self.left_motor.setVelocity(0.2 * speed[0])
            self.right_motor.setVelocity(0.2 * speed[1])


controller = Controller()
controller.run()

File: projects/samples/devices/controllers/distance_sensor/distance_sensor.c
/*
 * Copyright 1996-2023 Cyberbotics Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 * Description:  An example of use of the distance sensors
 */

#include <webots/distance_sensor.h>
#include <webots/motor.h>
#include <webots/robot.h>

#define NB_SENSORS 8
#define TIME_STEP 64
#define RANGE (1024 / 2)

int main(void) {
  int i, j;
  WbDeviceTag ps[NB_SENSORS];
  char name[] = "ds0";
  double speed[2];
  double sensor_value[NB_SENSORS];
  double matrix[2][NB_SENSORS] = {{11, 12, 8, -2, -3, -5, -7, -9}, {-9, -8, -5, -1, -2, 6, 12, 11}};

  wb_robot_init();

  /* get handlers to and enable the distance sensor devices */
  for (i = 0; i < NB_SENSORS; i++) {
    ps[i] = wb_robot_get_device(name); /* get a handler to the sensor */
    /* Perform distance measurements every TIME_STEP millisecond. */
    wb_distance_sensor_enable(ps[i], TIME_STEP);
    /* Increase the device name to "ir1", "ir2", etc. */
    name[2]++;
  }

  /* get a handler to the motors and set target position to infinity (speed control). */
  WbDeviceTag left_motor = wb_robot_get_device("left wheel motor");
  WbDeviceTag right_motor = wb_robot_get_device("right wheel motor");
  wb_motor_set_position(left_motor, INFINITY);
  wb_motor_set_position(right_motor, INFINITY);
  wb_motor_set_velocity(left_motor, 0.0);
  wb_motor_set_velocity(right_motor, 0.0);

  /* control loop */
  while (wb_robot_step(TIME_STEP) != -1) {
    for (i = 0; i < NB_SENSORS; i++)
      sensor_value[i] = wb_distance_sensor_get_value(ps[i]);
    for (i = 0; i < 2; i++) {
      speed[i] = 0;
      for (j = 0; j < NB_SENSORS; j++) {
        /*
         * We need to recenter the value of the sensor to be able to get
         * negative values too. This will allow the wheels to go
         * backward too.
         */
        speed[i] += matrix[i][j] * (1 - (sensor_value[j] / RANGE));
      }
    }

    /* Set the motor speeds */
    wb_motor_set_velocity(left_motor, 0.2 * speed[0]);
    wb_motor_set_velocity(right_motor, 0.2 * speed[1]);
  }

  wb_robot_cleanup();

  return 0;
}
